src/server/transaction.cc

```diff
@@ -1298,14 +1278,24 @@ void Transaction::ExpireShardCb(ArgSlice wkeys, EngineShard* shard) {
   CHECK_GE(DecreaseRunCnt(), 1u);
 }
 
-OpStatus Transaction::RunSquashedMultiCb(RunnableType cb) {
+void Transaction::RunSquashedMultiCb(RunnableType cb) {
   DCHECK(multi_ && multi_->role == SQUASHED_STUB);
   DCHECK_EQ(unique_shard_cnt_, 1u);
+  SafeRunCb(cb);
   auto* shard = EngineShard::tlocal();
-  auto status = cb(this, shard);
   shard->db_slice().OnCbFinish();
   LogAutoJournalOnShard(shard);
-  return status;
+}
+
+void Transaction::SafeRunCb(RunnableType cb) {
+  try {
+    local_result_ = cb(this, EngineShard::tlocal());
+  } catch (std::bad_alloc&) {
+    LOG_FIRST_N(ERROR, 16) << " out of memory";
+    local_result_ = OpStatus::OUT_OF_MEMORY;
+  } catch (std::exception& e) {
+    LOG(FATAL) << "Unexpected exception " << e.what();
+  }
```

Comment:
This is not safe as you call it from RunInShard() and write into local_result_ unconditionally

Note that RunInShard() doesn't write into local result with an OK status, it does so only for out of memory... And I actually don't understand why we do it either? https://github.com/dragonflydb/dragonfly/pull/2393#discussion_r1450346061

Reply:
Good catch! Fixed.